spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java

```diff
@@ -150,8 +167,54 @@ else if (reply instanceof AbstractIntegrationMessageBuilder) {
 			}
 		}
 
-		Message<?> replyMessage = createOutputMessage(reply, requestHeaders);
-		sendOutput(replyMessage, replyChannel);
+		if (this.asyncReplySupported && reply instanceof ListenableFuture<?>) {
+			ListenableFuture<?> future = (ListenableFuture<?>) reply;
+			final Object theReplyChannel = replyChannel;
+			future.addCallback(new ListenableFutureCallback<Object>() {
+
+				@Override
+				public void onSuccess(Object result) {
+					try {
+						sendOutput(createOutputMessage(result, requestHeaders), theReplyChannel, false);
+					}
+					catch (Exception e) {
+						Exception exceptionToLog = e;
+						if (!(e instanceof MessagingException)) {
+							exceptionToLog = new MessageHandlingException(requestMessage, e);
+						}
+						logger.error("Failed to send async reply: " + result.toString(), exceptionToLog);
```

Comment:
I don't want to call `onFailure()` because failures there will be misleading `Async exception received` would be incorrect, we received a good result, but couldn't sent it. I agree that we could send a failure here to the error channel but we'd need proper error messages if the failure is due to this case.

I am not sure what you are referring to, with regards to application events.

I'll take a look tomorrow unless you polish it first :smile: